extension/observer/dockerobserver/extension.go

```diff
@@ -36,6 +43,140 @@ func (d *dockerObserver) Shutdown(ctx context.Context) error {
 	return nil
 }
 
+// endpointsForContainer generates a list of observer.Endpoint given a Docker ContainerJSON.
+// This function will only generate endpoints if a container is in the Running state and not Paused.
+func (d *dockerObserver) endpointsForContainer(c *dtypes.ContainerJSON) []observer.Endpoint {
+	cEndpoints := make([]observer.Endpoint, 0)
+
+	if c.State.Running && !c.State.Paused {
+		knownPorts := map[nat.Port]bool{}
+		for k := range c.Config.ExposedPorts {
+			knownPorts[k] = true
+		}
+
+		// iterate over exposed ports and try to create endpoints
+		for portObj := range knownPorts {
+			endpoint := d.endpointForPort(portObj, c)
+			// the endpoint was not set, so we'll drop it
+			if endpoint == (observer.Endpoint{}) {
+				continue
+			}
+			cEndpoints = append(cEndpoints, endpoint)
+		}
+	}
+
+	for _, e := range cEndpoints {
+		s, _ := json.MarshalIndent(e, "", "\t")
+		d.logger.Debug(fmt.Sprintf("Discovered Docker container endpoint:\n%v\n", string(s)))
+	}
+
+	return cEndpoints
+}
+
+// endpointForPort creates an observer.Endpoint for a given port that is exposed in a Docker container.
+// Each endpoint has a unique ID generated by the combination of the container.ID, container.Name,
+// underlying host name, and the port number.
+// Uses the user provided config settings to override certain fields.
+func (d *dockerObserver) endpointForPort(portObj nat.Port, c *dtypes.ContainerJSON) observer.Endpoint {
+	endpoint := observer.Endpoint{}
+	port := portObj.Int()
+	proto := portObj.Proto()
+
+	mappedPort, mappedIP := findHostMappedPort(c, portObj)
+	if d.config.IgnoreNonHostBindings && mappedPort == 0 && mappedIP == "" {
+		return endpoint
+	}
+
+	// unique ID per host/container/port
+	// (docker_observer)container_ID-container_name-host_name:port
+	id := observer.EndpointID(
+		fmt.Sprintf(
+			"(%s)%s-%s-%s:%d",
+			typeStr, c.ID, c.Name, mappedIP, mappedPort,
+		),
+	)
+
+	details := &observer.Container{
+		Name:        c.Name,
+		Image:       c.Config.Image,
+		Command:     strings.Join(c.Config.Cmd, " "),
+		ContainerID: c.ID,
+		Transport:   portProtoToTransport(proto),
+		Labels:      c.Config.Labels,
+	}
+	var target string
+
+	// Set our target and hostname based on config settings
+	if d.config.UseHostnameIfPresent && c.Config.Hostname != "" {
+		target = c.Config.Hostname
+		details.Host = c.Config.Hostname
+	} else {
+		// Use the IP Address of the first network we iterate over.
+		// This can be made configurable if so desired.
+		for _, n := range c.NetworkSettings.Networks {
+			target = n.IPAddress
+			details.Host = n.IPAddress
+			break
+		}
+
+		// If we still haven't gotten a host at this point and we are using
+		// host bindings, just make it localhost.
+		if target == "" && d.config.UseHostBindings {
+			target = "127.0.0.1"
+			details.Host = "127.0.0.1"
+		}
+	}
+
+	// If we are using HostBindings & port and IP are set, use those
+	if d.config.UseHostBindings && mappedPort != 0 && mappedIP != "" {
```

Comment:
Should this happen before the hostname/network target logic since it overwrites their values?

Reply:
I'm not sure what the the ideal is here since both config options, when enabled, modify the target/host. Maybe just documenting one takes precedence over the other in the config would suffice?